src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs

```diff
@@ -600,35 +612,67 @@ internal override object UnMarshal(CLRStackValue dsObject, System.Type expectedC
             ICollection collection = null;
             if (dsObject.IsEnumerable)
             {
+                // collection could be T[] or List<object>.
                 collection = ToICollection(dsObject, arrayType, runtimeCore);
             }
             else
             {
                 // If dsObject is non array pointer but the expectedCLRType is IEnumerable, promote the dsObject to a collection.
                 Validity.Assert(typeof(IEnumerable).IsAssignableFrom(expectedCLRType));
+                
                 var obj = primitiveMarshaler.UnMarshal(dsObject, elementType, runtimeCore);
-                collection = new ArrayList(new object[] { obj });
+                collection = new List<object> { obj };
             }
 
             if (expectedCLRType.IsGenericType && !expectedCLRType.IsInterface)
             {
                 if (!collection.GetType().IsArray)
                 {
-                    Validity.Assert(collection is ArrayList);
-                    collection = (collection as ArrayList).ToArray(elementType);
+                    collection = (collection as List<object>).ToArray();
                 }
                 return Activator.CreateInstance(expectedCLRType, new[] { collection });
             }
 
             if (expectedCLRType.IsArray || expectedCLRType.IsGenericType)
             {
-                var list = collection as ArrayList;
-                if (null != list)
-                    return list.ToArray(elementType);
-            }
+                if (collection is Array) return collection;
 
+                if(elementType == typeof(int))
+                {
+                    return ConvertToTypedArray<int>(collection);
+                }
+                if (elementType == typeof(long))
+                {
+                    return ConvertToTypedArray<long>(collection);
+                }
+                if (elementType == typeof(double))
+                {
+                    return ConvertToTypedArray<double>(collection);
```

Comment:
another alternative is to use collection.Cast<T>().ToArray(); instead of the custom `ConvertToTypedArray` function you implemented.